### PR TITLE
[NPM-2758] Create System Probe Liveness Check

### PR DESCRIPTION
**Pull request by tblake333**
### What does this PR do?

Expose a new endpoint for system probe (5558 suggested to avoid conflict with other health ports). This endpoint may be used for Kubernetes liveness checks.

Endpoint value can be changed through the `system-probe.yaml` file:

```yaml
# system-probe.yaml
...
system_probe_config:
    ...
    health_port: <port>
    ...
...
```

### Additional Notes
This PR introduces the health check, but only implements the check for the network_tracer module. The other modules in `system-probe` should be supported in the future.

Helm chart updates to be made in a later PR.

### Describe how to test/QA your changes

To test the endpoint:
1. Build `system-probe` from an `agent-dev_box` VM
```bash
cd /git/datadog-agent
inv -e system-probe.build
```
2. Run `system-probe`
```bash
sudo bin/system-probe/system-probe
```
3. In another terminal of the `agent-dev_box` VM, hit the endpoint:
```bash
curl localhost:5558/health
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/config/config_template.yaml
Comment (jhgilbert) [inline pkg/config/config_template.yaml:1756]:
```suggestion
  ## Default is 0 (disabled). Set a valid port number (example: 5556) to enable.
```

File: pkg/config/config_template.yaml
Comment (jhgilbert) [inline pkg/config/config_template.yaml:1754]:
```suggestion
  ## The Agent can expose its health check on a dedicated HTTP port.
```

File: pkg/config/config_template.yaml
Comment (jhgilbert) [inline pkg/config/config_template.yaml:1755]:
```suggestion
  ## This is useful for orchestrators that support HTTP probes.
```

**Review by jhgilbert — APPROVED** [review]
Approving with a few minor suggestions for style. Thanks!

File: releasenotes/notes/system-probe-health-check-efd20595e41b4e57.yaml
Comment (hmahmood) [inline releasenotes/notes/system-probe-health-check-efd20595e41b4e57.yaml:5]: Was there more here?

File: cmd/system-probe/subcommands/run/command.go
Comment (hmahmood) [inline cmd/system-probe/subcommands/run/command.go:249]: You want to do `%s` here.

File: cmd/system-probe/subcommands/run/command.go
Comment (hmahmood) [inline cmd/system-probe/subcommands/run/command.go:251]: This can be an `info` level log.

**Review by ogaca-dd — APPROVED** [review]

File: releasenotes/notes/system-probe-health-check-efd20595e41b4e57.yaml
Comment (tblake333) [reply releasenotes/notes/system-probe-health-check-efd20595e41b4e57.yaml:5]: There was... Thanks for catching this!

File: cmd/system-probe/api/module/loader.go
Comment (brycekahle) [inline cmd/system-probe/api/module/loader.go:180]: The health check cannot error on register, so I'm not sure this log is valuable.

File: cmd/system-probe/api/module/loader.go
Comment (brycekahle) [inline cmd/system-probe/api/module/loader.go:206]: A couple concerns here:
1. According to the health check docs, we need to read from this channel *at least* every 15 seconds. Since this loop is blocked by a 15 second tick (`now = <-ticker.C` below), this is basically guaranteed to be never true.
2. This is not how they recommend to read from this channel. They suggest to use a `select` where the health channel is one option, and your work is the other. See other usages of the package for how other components are using it.
3. This health check doesn't actually detect if there is a critical problem with any of the system-probe modules, because it isn't in the critical path. The point of the health check is to detect when the component is no longer processing input. What that means, is definitely module-specific, but the `GetStats` function is definitely not in the critical path. We would probably need each module to register their own liveness checks.

File: cmd/system-probe/subcommands/run/command.go
Comment (brycekahle) [inline cmd/system-probe/subcommands/run/command.go:249]: nit:
```suggestion
			return log.Errorf("error starting health check server, exiting: %s", err)
```

File: cmd/system-probe/subcommands/run/command.go
Comment (brycekahle) [inline cmd/system-probe/subcommands/run/command.go:251]: nit:
```suggestion
		log.Infof("health check server listening on port %d", healthPort)
```

File: cmd/system-probe/subcommands/run/command.go
Comment (brycekahle) [inline cmd/system-probe/subcommands/run/command.go:245]: Can you add a `HealthPort int` to the config object in `cmd/system-probe/config/config.go` and populate that in `load()` function of the same file? Then use `cfg.HealthPort` here instead.

File: Dockerfiles/manifests/all-containers/daemonset.yaml
Comment (brycekahle) [inline Dockerfiles/manifests/all-containers/daemonset.yaml:1]: I don't think you should manually update this file. It is auto-generated from the public helm chart. You will probably need to open a PR once the helm chart is updated.

Comment (brycekahle) [reply Dockerfiles/manifests/all-containers/daemonset.yaml:1]: See `Dockerfiles/manifests/README.md` for details

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: cmd/system-probe/api/module/loader.go
Comment (hmahmood) [reply cmd/system-probe/api/module/loader.go:206]: For (3.), doesn't the agent status hit the stats endpoint? So this would be roughly equivalent to that. The work to add a liveness check for each module is likely much more involved, and we can do a follow up for that.

Comment (brycekahle) [reply cmd/system-probe/api/module/loader.go:206]: `GetStats` can continue to return success, even if NPM is not processing any from the ebpf side. That is what I mean by not being in the critical path.

Comment (hmahmood) [reply cmd/system-probe/api/module/loader.go:206]: There are some tricky questions if we did per-module health checks. e.g., what do we report if one module has issues? Is the whole system-probe health check bad at that point? Module health can also be a complex thing; several things in network_tracer could indicate ill health.

I think at least we can reflect what agent status shows for now. We should open a new ticket for the per-module health check.

Comment (brycekahle) [reply cmd/system-probe/api/module/loader.go:206]: From my understanding, the point of a liveness check is to know when the container needs to be restarted.

**Review by brycekahle — APPROVED** [review]

**Review by hmahmood — APPROVED** [review]

**Pull request merged**
